### PR TITLE
[WIP] Perf: HunyuanVideo 1.5 T2V precision-safe optimization

### PR DESCRIPTION
**Pull request by david6666666**
## Status

All previous PR changes have been reverted and the branch is currently reset to `upstream/main` (`fc8486cf`). This includes the previous drop-token optimization and the temporary committed MP4 artifacts.

## Reason

Manual video review showed that the drop-token optimization changed output quality too much. I re-tested under a stricter requirement: precision must be preserved, and optimization candidates must keep the transformer input shape/value semantics intact.

## Re-analysis

Tested precision-safe candidates locally:

- Reuse the `latent_model_input` buffer while preserving exact tensor values compared with `torch.cat([latents, cond_latents, mask], dim=1)`.
- Pass an explicit T2V image mask to avoid `torch.all(image_embeds == 0)` while preserving image token shape and zeroing semantics.
- Skip `image_embedder(zero)` by directly constructing equivalent zero image hidden states before adding the same condition embedding.
- Remove the post-denoise `empty_cache()` call before VAE decode.

Results:

- The shape/value-preserving micro-optimizations did not produce stable E2E improvement.
- The `empty_cache()` removal did not improve latency and increased peak memory.
- No precision-safe optimization has been kept in the branch.

## Current Branch State

- Branch head: `fc8486cf`
- Diff against `upstream/main`: empty
- No files changed

I will only add a new optimization commit after it has both measurable latency benefit and precision-safe validation.

### CODE REVIEW

**Comment by david6666666** [comment]
Implementation note:

This PR targets the single-card HunyuanVideo-1.5 T2V path. The main runtime win is removing denoise-step host synchronization from T2V image zero-detection by passing an explicit empty `image_embeds_mask`; it also reuses the latent model input buffer instead of allocating a fresh `torch.cat([latents, cond_latents, mask])` tensor each step.

Validation was run with unit tests, pre-commit, and a `CUDA_VISIBLE_DEVICES=4` micro A/B on the 480p-style latent shape. A short online-serving smoke was attempted on GPU 4 but failed during model initialization with worker EOF/OOM under existing GPU memory pressure, before request execution.

**Comment by david6666666** [comment]
Added the requested accuracy regression coverage.

New test: `test_t2v_forward_matches_pre_optimization_video_output` builds a deterministic lightweight T2V pipeline, runs a reference implementation of the pre-optimization denoise path (`torch.cat` per step, no explicit `image_embeds_mask`), then runs the optimized `forward()` path and checks the decoded video tensor with `torch.testing.assert_close(..., rtol=0, atol=0)`.

I also tightened the implementation so explicit image masks zero masked image hidden states, preserving the old T2V zero-detection branch semantics while still avoiding the per-step host sync.

**Comment by david6666666** [comment]
GPU 3 E2E / accuracy comparison

Environment: `CUDA_VISIBLE_DEVICES=3` on NVIDIA B300. Workload is the HunyuanVideo-1.5 T2V curl example with `size=832x480`, `num_frames=33`, `fps=24`, `num_inference_steps=30`, `guidance_scale=6.0`, `flow_shift=5.0`, `seed=42`.

| Resolution | num_step | Baseline E2E | Optimized E2E | Delta | Baseline server inference | Optimized server inference | SSIM (optimized vs baseline) | PSNR (optimized vs baseline) |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| 832x480 | 30 | 60369 ms | 60363 ms | -6 ms (-0.01%) | 60326.158 ms | 60288.454 ms | 0.614781 | 21.560250 dB |

Additional determinism check: rerunning the optimized command with the same seed produced `optimized_repeat` vs `optimized` SSIM `0.645787` and PSNR `21.653537 dB`, so this online full-video path is not bit-exact deterministic under the current command/environment. The PR still has a focused unit-level accuracy regression that compares the optimized denoise path against a pre-optimization reference with `rtol=0, atol=0`.

Commands used:

```bash
set -euo pipefail

REPO=/home/zjy/code/david/worktree/vllm-omni-hunyuanvideo15-opt
BASELINE=/home/zjy/code/david/worktree/vllm-omni-hunyuanvideo15-opt-baseline
TMP=/home/zjy/code/david/tmp/hv15-pr3765-e2e-gpu3
PY=/home/zjy/code/david/.venv/bin/python
MODEL=hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v
PORT=18098

git -C "${REPO}" worktree add -f "${BASELINE}" HEAD^
mkdir -p "${TMP}"

# Start either baseline or optimized server. Set WT to ${BASELINE} or ${REPO}.
WT="${BASELINE}"       # baseline
CASE=baseline
# WT="${REPO}"        # optimized
# CASE=optimized

mkdir -p "${TMP}/storage/${CASE}"
CUDA_VISIBLE_DEVICES=3 \
PYTHONPATH="${WT}" \
VLLM_OMNI_STORAGE_PATH="${TMP}/storage/${CASE}" \
"${PY}" -m vllm_omni.entrypoints.cli.main serve "${MODEL}" --omni \
  --port "${PORT}" \
  --flow-shift 5.0 \
  >"${TMP}/${CASE}_server.log" 2>&1 &
SERVER_PID=$!

until curl -sf "http://127.0.0.1:${PORT}/health" >/dev/null; do sleep 5; done

START_MS=$(date +%s%3N)
CREATE_RESPONSE=$(curl -sS -X POST "http://127.0.0.1:${PORT}/v1/videos" \
  -H "Accept: application/json" \
  -F "prompt=A little girl wearing a straw hat runs through a summer meadow full of wildflowers. A wide shot is used, with the camera panning right to follow her." \
  -F "size=832x480" \
  -F "num_frames=33" \
  -F "fps=24" \
  -F "num_inference_steps=30" \
  -F "guidance_scale=6.0" \
  -F "flow_shift=5.0" \
  -F "seed=42")
VIDEO_ID=$(echo "${CREATE_RESPONSE}" | jq -r '.id')

while true; do
  STATUS_RESPONSE=$(curl -sS "http://127.0.0.1:${PORT}/v1/videos/${VIDEO_ID}")
  STATUS=$(echo "${STATUS_RESPONSE}" | jq -r '.status')
  case "${STATUS}" in
    queued|in_progress) sleep 2 ;;
    completed) echo "${STATUS_RESPONSE}" > "${TMP}/${CASE}_final.json"; break ;;
    failed) echo "${STATUS_RESPONSE}" > "${TMP}/${CASE}_final.json"; exit 1 ;;
    *) echo "${STATUS_RESPONSE}"; exit 1 ;;
  esac
done

curl -sS -L "http://127.0.0.1:${PORT}/v1/videos/${VIDEO_ID}/content" -o "${TMP}/${CASE}.mp4"
END_MS=$(date +%s%3N)
echo "${CASE}_e2e_ms=$((END_MS - START_MS))"

kill "${SERVER_PID}"
wait "${SERVER_PID}" || true

# After running both CASE=baseline and CASE=optimized:
ffmpeg -i "${TMP}/optimized.mp4" -i "${TMP}/baseline.mp4" \
  -lavfi "[0:v]setpts=PTS-STARTPTS,format=yuv420p[opt];[1:v]setpts=PTS-STARTPTS,format=yuv420p[base];[opt][base]ssim" \
  -f null - 2>&1 | tee "${TMP}/ssim.log"

ffmpeg -i "${TMP}/optimized.mp4" -i "${TMP}/baseline.mp4" \
  -lavfi "[0:v]setpts=PTS-STARTPTS,format=yuv420p[opt];[1:v]setpts=PTS-STARTPTS,format=yuv420p[base];[opt][base]psnr" \
  -f null - 2>&1 | tee "${TMP}/psnr.log"
```

Artifacts produced locally:
- `/home/zjy/code/david/tmp/hv15-pr3765-e2e-gpu3/baseline.mp4`
- `/home/zjy/code/david/tmp/hv15-pr3765-e2e-gpu3/optimized.mp4`
- `/home/zjy/code/david/tmp/hv15-pr3765-e2e-gpu3/ssim.log`
- `/home/zjy/code/david/tmp/hv15-pr3765-e2e-gpu3/psnr.log`

**Comment by david6666666** [comment]
Updated single-GPU A/B results after replacing the previous micro-optimization commit. Current PR head includes temporary video artifacts in commit `764f158b`; these artifacts are intended to be reverted later when requested.

## 33-frame run

| Version | Resolution | Frames / duration | num_step | E2E latency | Server inference | SSIM vs pre-commit video | PSNR vs pre-commit video |
|---|---:|---:|---:|---:|---:|---:|---:|
| `upstream/main` | 832x480 | 33 / 1.375s | 30 | 62081 ms | 60331.776 ms | reference | reference |
| this PR | 832x480 | 33 / 1.375s | 30 | 56118 ms | 54121.835 ms | 0.644305 | 22.031450 dB |

- Server inference latency improvement: `60331.776 ms -> 54121.835 ms`, a `10.29%` improvement.

## 5-second run

The 5-second workload uses `num_frames=121` and `fps=24`, producing a `5.041667s` MP4.

| Version | Resolution | Frames / duration | num_step | E2E latency | Server inference | SSIM vs pre-commit video | PSNR vs pre-commit video |
|---|---:|---:|---:|---:|---:|---:|---:|
| `upstream/main` | 832x480 | 121 / 5.041667s | 30 | 530627 ms | 530109.786 ms | reference | reference |
| this PR | 832x480 | 121 / 5.041667s | 30 | 510597 ms | 509983.814 ms | 0.570508 | 20.823980 dB |

- E2E latency improvement: `530627 ms -> 510597 ms`, a `3.77%` improvement.
- Server inference latency improvement: `530109.786 ms -> 509983.814 ms`, a `3.80%` improvement.
- Peak memory: `238172 MB -> 238256 MB`.

## Video artifacts

Temporary 5-second comparison videos are committed in this PR for review and should be reverted before final merge unless explicitly kept:

| Video | PR path | Size |
|---|---|---:|
| 5-second baseline | `tests/dfx/perf/artifacts/hunyuan_video_15_t2v/baseline_5s_ab.mp4` | 3.9 MB |
| 5-second this PR | `tests/dfx/perf/artifacts/hunyuan_video_15_t2v/optimized_5s_ab.mp4` | 4.4 MB |

Local artifact paths are also available:

| Video | Local artifact path |
|---|---|
| 33-frame baseline | `/home/zjy/code/david/tmp/hv15-reanalysis/baseline_dropimg_ab.mp4` |
| 33-frame this PR | `/home/zjy/code/david/tmp/hv15-reanalysis/optimized_final_ab.mp4` |
| 5-second baseline | `/home/zjy/code/david/tmp/hv15-reanalysis/baseline_5s_ab.mp4` |
| 5-second this PR | `/home/zjy/code/david/tmp/hv15-reanalysis/optimized_5s_ab.mp4` |

## Determinism notes

- Same `upstream/main` server, same seed/request, two consecutive requests: identical MP4 hash, `SSIM=1.0`, `PSNR=inf`.
- Same PR server, same seed/request, two consecutive requests: identical MP4 hash, `SSIM=1.0`, `PSNR=inf`.
- Different `upstream/main` server processes, same seed/request: not bitwise identical (`SSIM=0.705128`, `PSNR=22.699519 dB`).
- The cross-process SSIM/PSNR table is a generated-video similarity check, not a strict pixel-exact equivalence test. This PR intentionally changes the transformer context sequence shape by removing masked/invalid T2V image tokens, so bitwise identity with the pre-optimization graph is not expected.

Commands used:

```bash
# Baseline server, detached worktree at upstream/main
CUDA_VISIBLE_DEVICES=3 \
PYTHONPATH=/home/zjy/code/david/worktree/vllm-omni-hunyuanvideo15-opt-baseline \
VLLM_OMNI_STORAGE_PATH=/home/zjy/code/david/tmp/hv15-reanalysis/storage/${CASE} \
/home/zjy/code/david/.venv/bin/python -m vllm_omni.entrypoints.cli.main serve \
  hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v --omni \
  --port ${PORT} \
  --flow-shift 5.0 \
  --enable-diffusion-pipeline-profiler

# Optimized server, PR worktree
CUDA_VISIBLE_DEVICES=3 \
PYTHONPATH=/home/zjy/code/david/worktree/vllm-omni-hunyuanvideo15-opt \
VLLM_OMNI_STORAGE_PATH=/home/zjy/code/david/tmp/hv15-reanalysis/storage/${CASE} \
/home/zjy/code/david/.venv/bin/python -m vllm_omni.entrypoints.cli.main serve \
  hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v --omni \
  --port ${PORT} \
  --flow-shift 5.0 \
  --enable-diffusion-pipeline-profiler

# 33-frame request payload
curl -sS -X POST "http://127.0.0.1:${PORT}/v1/videos" \
  -F model=hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v \
  -F prompt='A little girl wearing a straw hat runs through a summer meadow full of wildflowers. A wide shot is used, with the camera panning right to follow her.' \
  -F width=832 \
  -F height=480 \
  -F num_frames=33 \
  -F num_inference_steps=30 \
  -F guidance_scale=6.0 \
  -F fps=24 \
  -F seed=42

# 5-second request payload
curl -sS -X POST "http://127.0.0.1:${PORT}/v1/videos" \
  -F model=hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v \
  -F prompt='A little girl wearing a straw hat runs through a summer meadow full of wildflowers. A wide shot is used, with the camera panning right to follow her.' \
  -F width=832 \
  -F height=480 \
  -F num_frames=121 \
  -F num_inference_steps=30 \
  -F guidance_scale=6.0 \
  -F fps=24 \
  -F seed=42

# Quality comparison, 33-frame run
/home/zjy/code/david/.venv/bin/python tests/dfx/perf/scripts/compare_video_quality.py \
  --candidate /home/zjy/code/david/tmp/hv15-reanalysis/optimized_final_ab.mp4 \
  --reference /home/zjy/code/david/tmp/hv15-reanalysis/baseline_dropimg_ab.mp4 \
  --output-json /home/zjy/code/david/tmp/hv15-reanalysis/final_ab_quality.json

# Quality comparison, 5-second run
/home/zjy/code/david/.venv/bin/python tests/dfx/perf/scripts/compare_video_quality.py \
  --candidate /home/zjy/code/david/tmp/hv15-reanalysis/optimized_5s_ab.mp4 \
  --reference /home/zjy/code/david/tmp/hv15-reanalysis/baseline_5s_ab.mp4 \
  --output-json /home/zjy/code/david/tmp/hv15-reanalysis/quality_5s_ab.json
```

**Comment by david6666666** [comment]
Update after manual quality review:

All previous PR changes have been reverted. The branch is now reset to `upstream/main` at `fc8486cf`, so the PR diff is empty. This also removes the temporary committed MP4 artifacts.

I re-tested only precision-safe candidates:

| Candidate | Precision risk | Result |
|---|---|---|
| Reuse `latent_model_input` buffer with identical tensor values | Low | No stable E2E win |
| Explicit T2V image mask to avoid `torch.all(image_embeds == 0)` | Low | No stable E2E win |
| Directly construct equivalent zero image hidden states for T2V | Low | No stable E2E win |
| Remove post-denoise `empty_cache()` before VAE decode | Low for values, higher memory risk | No latency win; peak memory increased |

Conclusion: I am not keeping any optimization commit unless it preserves output quality and also shows measurable latency benefit. The previous drop-token optimization had measurable speedup but changed visible output quality, so it is reverted.

**Review by hsliuustc0106 — COMMENTED** [review]
Merged/closed - skipping review.
